File: c_ds/train/basics/merge_intervals.py
class Solution:
    def merge(self, intervals: list[list[int]]) -> list[list[int]]:
        intervals.sort(key = lambda x : x[0])
        result = [intervals[0]]
        for item in intervals[1:]:
            merged_item = result[-1]
            if  merged_item[1] < item[0]:
                result.append(item)
            else:
                merged_item[1] = max(merged_item[1], item[1])
        return result




def test():
    # intervals = [[1,3],[2,6],[8,10],[15,18]]
    intervals = [[2,6],[1,3],[1,4],[8,10],[15,18]]
    s = Solution()
    final_list = s.merge(intervals)
    print(f"\t for input:{intervals} \n\t merged:{final_list}")



if __name__ == "__main__":
    test()

File: c_ds/train/basics/merge_intervals.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

template <typename Stream, typename T>
Stream& operator<< (Stream& stream, vector<T> elements){
    stream << "[";
    for ( auto element : elements){
        stream << element << ", ";
    }
    stream << "] ";
    return stream;
}


class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        cout << "\t inputs:" << intervals << endl;
        vector<vector<int>> result;
        std::sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b){return a[0] < b[0];});
        cout << "\t sorted inputs:" << intervals << endl;
        for ( auto interval : intervals){
            if ( result.size() == 0){
                result.push_back(interval);
            } else {
                cout << "\t\t -- " << result[result.size()-1] << " -- interval:" << interval << endl;
                if (result[result.size()-1][1] >= interval[0]){
                    result[result.size()-1][1] = std::max(result[result.size()-1][1], interval[1]);
                } else {
                    result.push_back(interval);
                }
            }
            cout << result << endl;
        }
        return result;
    }
};
void test(){
    // vector<vector<int>> intervals = [[1,3],[2,6],[8,10],[15,18]];
    // vector<vector<int>> intervals = {{2,6},{1,3},{1,4},{8,10},{15,18}};
    vector<vector<int>> intervals = {{1,4},{4,5}};
    Solution s;
    auto final_list = s.merge(intervals);
    cout << "\t for input:" << intervals <<  "\n\t merged:" << final_list << endl;
}


int main(){
    test();
    return 0;
}